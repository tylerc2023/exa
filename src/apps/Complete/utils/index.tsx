export * from './configs';
export * from './useKeyboardHeight';
export * from './userFactory';
