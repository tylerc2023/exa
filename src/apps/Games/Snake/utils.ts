export type BoardObject = Array<Array<0 | 1 | 2>>;

export const generateRandom = (size: number) =>
  Math.floor(Math.random() * size);

export const addFood = (
  board: BoardObject,
  used: {[key: string]: boolean} = {},
): boolean => {
  const x = generateRandom(board.length);
  const y = generateRandom(board.length);

  const combo = `${x}${y}`;
  if (Object.keys(used).length === board.length) {
    return false;
  }
  if (combo in used || board[x][y] !== 0) {
    used[combo] = true;
    addFood(board, used);
    return false;
  }
  board[y][x] = 2;
  return true;
};

export const addStarting = (board: BoardObject) => {
  const center = Math.floor(board.length / 2);
  board[center][center] = 1;
};

export const generateBoard = (size: number) => {
  const board: BoardObject = [];
  for (let i = 0; i < size; i++) {
    board[i] = [];
    for (let j = 0; j < size; j++) {
      board[i][j] = 0;
    }
  }

  addStarting(board);
  addFood(board);
  return board;
};

export const nextSnakePosition = () => undefined;
export const collision = (next: any) => next && false;
export const endGame = () => undefined;
export const eat = () => undefined;
