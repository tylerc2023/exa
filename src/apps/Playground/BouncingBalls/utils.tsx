import {Animated} from 'react-native';
import {State} from './BouncingBall';
import {CanvasDimensions} from './Canvas';

const getRandomNumber = (min: number, max: number): number =>
  Math.random() * (max - min) + min;

type GetSpeedProps = {
  state: State;
  canvas: CanvasDimensions;
  diameter: number;
  pulse: boolean;
};

export const getSpeed = ({state, canvas, diameter, pulse}: GetSpeedProps) => {
  const dMax = state.diameter >= diameter * 1.1;
  const dMin = state.diameter <= diameter * 0.9;
  // TODO: recenter so pulsing happens from middle
  const growing = dMax || dMin ? !state.growing : state.growing;
  const d = state.diameter + (!pulse ? 0 : growing ? 1 : -1);
  const xMin = canvas.x;
  const xMax = canvas.x + canvas.width;
  const yMin = canvas.y;
  const yMax = canvas.y + canvas.height;
  const xLarger = state.x + d >= xMax;
  const xSmaller = state.x <= xMin;
  const yLarger = state.y + d >= yMax;
  const ySmaller = state.y <= yMin;
  // TODO: add collision
  const xSpeed = xLarger || xSmaller ? -state.xSpeed : state.xSpeed;
  const ySpeed = yLarger || ySmaller ? -state.ySpeed : state.ySpeed;
  const x = xLarger ? xMax - d : xSmaller ? xMin : state.x + xSpeed;
  const y = yLarger ? yMax - d : ySmaller ? yMin : state.y + ySpeed;
  const dx = x + xSpeed;
  const dy = y + ySpeed;
  return {xSpeed, ySpeed, x: dx, y: dy, d, growing};
};

type GetInitialStateProps = {
  canvas: CanvasDimensions;
  diameter: number;
  speed: number;
};

export const getInitialState = ({
  canvas,
  diameter,
  speed,
}: GetInitialStateProps): State => {
  const starting = {
    x: getRandomNumber(canvas.x, canvas.x + canvas.width - diameter),
    y: getRandomNumber(canvas.y, canvas.y + canvas.height - diameter),
  };
  const growing = !!Math.floor(Math.random());
  return {
    ...starting,
    position: new Animated.ValueXY(starting),
    xSpeed: getRandomNumber(-speed, speed),
    ySpeed: getRandomNumber(-speed, speed),
    diameter,
    growing,
  };
};
