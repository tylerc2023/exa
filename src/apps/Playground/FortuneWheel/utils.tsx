type GetRandomRange = {
  min: number;
  max: number;
};

const getRandomRange = ({min, max}: GetRandomRange) => {
  return Math.floor(Math.random() * (max - min) + min);
};

type GetNewLocation = {
  minSpin: number;
  maxSpin: number;
  numOfSegments: number;
  location: number;
};

export const getNewLocation = ({
  minSpin,
  maxSpin,
  numOfSegments,
  location,
}: GetNewLocation) => {
  const spins = getRandomRange({min: minSpin, max: maxSpin}) * 360;
  const winner = Math.floor(Math.random() * numOfSegments);
  const winnerSection = Math.floor(winner * (360 / numOfSegments));
  const newLocation = location + spins + winnerSection;
  return newLocation;
};

type GetWinnerIndex = {
  location: number;
  numOfSegments: number;
  angleOfSegment: number;
};

export const getWinnerIndex = ({
  location,
  numOfSegments,
  angleOfSegment,
}: GetWinnerIndex) => {
  const deg = Math.abs(Math.round(location % 360));
  const index =
    (numOfSegments - Math.floor(deg / angleOfSegment)) % numOfSegments;
  return index;
};
