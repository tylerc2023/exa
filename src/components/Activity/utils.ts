import axios from 'axios';
import endOfWeek from 'date-fns/endOfWeek';
import format from 'date-fns/format';
import isToday from 'date-fns/isToday';
import startOfWeek from 'date-fns/startOfWeek';
import sub from 'date-fns/sub';
import {ActivityMatrix, Site} from './types';

type ApiResponse = {[unix: string]: number};

type ApiPromise = Promise<ApiResponse>;

type ApiInput = {
  username: string;
  site: Site;
};

type ActivitySquares = {
  matrix: ActivityMatrix;
  max: number;
  total: number;
};

export const getDateFormat = (date: number): string =>
  format(date, 'yyyy-MM-dd');

export const getSubmissionFormat = (count: number, date: number): string => {
  const submissions = count === 1 ? 'submission' : 'submissions';
  const day = isToday(date) ? 'today' : `on ${format(date, 'MMM dd, yyyy')}`;
  return `${count} ${submissions} ${day}`;
};

export const getActivitySquares = (): ActivitySquares => {
  const today = Date.now();
  const matrix = [];
  const oneDay = 60 * 60 * 24 * 1000;
  const begin = sub(startOfWeek(today), {years: 1}).valueOf();
  let end = endOfWeek(today).valueOf();
  let day = 0;
  let week = [];
  while (end >= begin) {
    week.unshift({date: end, count: 0});
    if (day > 5) {
      matrix.push(week);
      week = [];
      day = 0;
    } else {
      day++;
    }
    end -= oneDay;
  }

  return {matrix, max: 0, total: 0};
};

export const updateActivitySquares = (
  squares: ActivitySquares,
  active: ApiResponse,
): ActivitySquares => {
  let max = 0;
  let total = 0;
  const matrix = squares.matrix.map((week) =>
    week.map((day) => {
      const count = active[getDateFormat(day.date)] || 0;
      max = Math.max(count, max);
      total += count;
      return {...day, count};
    }),
  );
  return {matrix, max, total};
};

const getGithubActivity = async (username: string): Promise<ApiResponse> => {
  const url = `https://github-contributions-json-api.now.sh/api?username=${username}`;
  const res = await axios.get(url);
  return res.data;
};

const getLeetCodeActivity = async (username: string) => {
  const url = `https://leetcode.com/api/user_submission_calendar/${username}`;
  const res = await axios.get(url);
  const data = JSON.parse(res.data);
  return Object.keys(data).reduce((total: ApiResponse, item) => {
    const day = getDateFormat(Number(item) * 1000);
    if (day in total) total[day] += data[item];
    else total[day] = data[item];

    return total;
  }, {});
};

const getHackerRankActivity = async (username: string) => {
  const url = `https://www.hackerrank.com/rest/hackers/${username}/submission_histories`;
  const res = await axios.get(url);
  return Object.keys(res.data).reduce((total: ApiResponse, item) => {
    total[item] = Number(res.data[item]);
    return total;
  }, {});
};

export const getApiActivity = ({username, site}: ApiInput): ApiPromise => {
  switch (site) {
    case 'github':
      return getGithubActivity(username);
    case 'leetCode':
      return getLeetCodeActivity(username);
    case 'hackerRank':
      return getHackerRankActivity(username);
    default:
      return Promise.resolve({});
  }
};
