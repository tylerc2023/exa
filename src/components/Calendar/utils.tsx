import { Dayjs } from "dayjs";

export type Day = {
  id: string;
  display: string;
  current: boolean;
  header: boolean;
};

export type CalendarMatrix = Day[][];

const getDay = (
  id: string,
  display: number | string,
  current = false,
  header = false
): Day => ({ id, display: String(display), current, header });

const getUnix = (today: Dayjs, month: number, day: number) =>
  String(
    today
      .add(month, "month")
      .date(day)
      .format("MM DD YYYY")
      .valueOf()
  );

export const getCalendarMatrix = (date: Dayjs): CalendarMatrix => {
  const days = ["Sun", "Mon", "Tue", "Wed", "Thu", "Fri", "Sat"];
  const firstDay = Number(date.startOf("month").format("d"));
  const maxDays = Number(date.endOf("month").format("D"));
  const prevMaxDays = Number(
    date
      .subtract(1, "month")
      .endOf("month")
      .format("D")
  );
  let dayCounter = 1;
  let prevDayCounter = prevMaxDays - firstDay + 1;
  let nextDayCounter = 1;
  const calendarMatrix: CalendarMatrix = [];
  calendarMatrix.push(days.map(day => getDay(day, day, false, true)));
  for (let row = 1; row < 7; row++) {
    calendarMatrix[row] = [];
    for (let col = 0; col < 7; col++) {
      calendarMatrix[row][col] =
        row === 1 && col < firstDay
          ? getDay(getUnix(date, -1, prevDayCounter), prevDayCounter++)
          : row > 1 && dayCounter > maxDays
          ? getDay(getUnix(date, 1, nextDayCounter), nextDayCounter++)
          : getDay(getUnix(date, 0, dayCounter), dayCounter++, true);
    }
  }
  return calendarMatrix;
};
