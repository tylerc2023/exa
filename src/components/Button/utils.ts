import {StyleSheet} from 'react-native';
import {colorWithOpacity, Theme} from '../../utils';
import {Color} from './../../models/Theme';
import {ButtonEmphasis} from './types';

export const getButtonColor = (
  colorScheme: Color,
  color: keyof Color,
): string => colorScheme[color];

interface StyleInterface {
  colorScheme: Color;
  color: string;
  emphasis: ButtonEmphasis;
  noPadding?: boolean;
  disable?: boolean;
}

export const getStyles = ({
  colorScheme,
  color,
  emphasis,
  noPadding,
  disable,
}: StyleInterface): any =>
  StyleSheet.create({
    center: {
      alignSelf: 'center',
    },
    container: {
      backgroundColor:
        disable && emphasis === 'high'
          ? colorWithOpacity(color, 0.38)
          : emphasis === 'high'
          ? color
          : 'transparent',
      borderColor:
        disable && emphasis === 'medium'
          ? colorWithOpacity(color, 0.38)
          : emphasis === 'medium'
          ? colorScheme.dark
          : 'transparent',
      borderRadius: Theme.padding.p01,
      borderWidth: 1,
      flexDirection: 'row',
      padding: noPadding ? Theme.padding.p00 : Theme.padding.p02,
      paddingHorizontal: noPadding ? Theme.padding.p00 : Theme.padding.p04,
    },
    invisible: {
      opacity: 0,
    },
    label: {
      height: Theme.padding.p05,
      justifyContent: 'flex-start',
      marginVertical: Theme.padding.p01,
      paddingHorizontal: 0,
    },
    right: {
      alignSelf: 'flex-end',
    },
    text: {
      color: disable
        ? colorWithOpacity(color, 0.38)
        : emphasis === 'high'
        ? colorScheme.background
        : color,
    },
  });
