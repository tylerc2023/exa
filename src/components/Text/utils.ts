import {Theme} from '../../utils';

export type FontType =
  | 'h1'
  | 'h2'
  | 'h3'
  | 'h4'
  | 'h5'
  | 'h6'
  | 'subtitle1'
  | 'subtitle2'
  | 'body1'
  | 'body2'
  | 'button'
  | 'caption'
  | 'overline';

export type FontEmphasis = 'high' | 'medium' | 'low';

export const getFontType = (type?: FontType) =>
  type === 'h1'
    ? Theme.fontSize.h1
    : type === 'h2'
    ? Theme.fontSize.h2
    : type === 'h3'
    ? Theme.fontSize.h3
    : type === 'h4'
    ? Theme.fontSize.h4
    : type === 'h5'
    ? Theme.fontSize.h5
    : type === 'h6'
    ? Theme.fontSize.h6
    : type === 'subtitle1'
    ? Theme.fontSize.subtitle1
    : type === 'subtitle2'
    ? Theme.fontSize.subtitle2
    : type === 'body1'
    ? Theme.fontSize.body1
    : type === 'body2'
    ? Theme.fontSize.body2
    : type === 'button'
    ? Theme.fontSize.button
    : type === 'caption'
    ? Theme.fontSize.caption
    : type === 'overline'
    ? Theme.fontSize.overline
    : Theme.fontSize.body2;

export const getTextEmphasis = (emphasis?: FontEmphasis) =>
  emphasis === 'high'
    ? 0.87
    : emphasis === 'medium'
    ? 0.6
    : emphasis === 'low'
    ? 0.38
    : 1.0;
