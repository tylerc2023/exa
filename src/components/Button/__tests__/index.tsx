import React from "react";
import { Button } from "..";
import { mockRenderer } from "../../../utils/Mock";

it("renders correctly", () => {
  const dom = mockRenderer(
    <Button title="hello" onPress={() => undefined} />
  ).toJSON();
  expect(dom).toMatchSnapshot();
});
it("renders contained", () => {
  const dom = mockRenderer(
    <Button contained title="hello" onPress={() => undefined} />
  ).toJSON();
  expect(dom).toMatchSnapshot();
});
it("renders contained disable", () => {
  const dom = mockRenderer(
    <Button contained disable title="hello" onPress={() => undefined} />
  ).toJSON();
  expect(dom).toMatchSnapshot();
});
it("renders outlined", () => {
  const dom = mockRenderer(
    <Button outlined title="hello" onPress={() => undefined} />
  ).toJSON();
  expect(dom).toMatchSnapshot();
});
it("renders fab", () => {
  const dom = mockRenderer(
    <Button fab title="hello" onPress={() => undefined} />
  ).toJSON();
  expect(dom).toMatchSnapshot();
});
it("renders text", () => {
  const dom = mockRenderer(
    <Button text title="hello" onPress={() => undefined} />
  ).toJSON();
  expect(dom).toMatchSnapshot();
});
it("renders sizing", () => {
  const dom = mockRenderer(
    <Button center half wrap disable title="hello" onPress={() => undefined} />
  ).toJSON();
  expect(dom).toMatchSnapshot();
});
it("renders icon", () => {
  const dom = mockRenderer(
    <Button icon="check" onPress={() => undefined} />
  ).toJSON();
  expect(dom).toMatchSnapshot();
});
it("renders icon title", () => {
  const dom = mockRenderer(
    <Button title="hello" icon="check" onPress={() => undefined} />
  ).toJSON();
  expect(dom).toMatchSnapshot();
});
