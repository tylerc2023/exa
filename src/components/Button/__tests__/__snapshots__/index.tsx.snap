// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders correctly 1`] = `
<View
  accessible={true}
  focusable={true}
  onClick={[Function]}
  onResponderGrant={[Function]}
  onResponderMove={[Function]}
  onResponderRelease={[Function]}
  onResponderTerminate={[Function]}
  onResponderTerminationRequest={[Function]}
  onStartShouldSetResponder={[Function]}
  style={
    Object {
      "alignItems": "center",
      "backgroundColor": "transparent",
      "borderColor": "transparent",
      "borderRadius": 4,
      "borderWidth": 1,
      "flexDirection": "row",
      "justifyContent": "center",
      "opacity": 1,
      "padding": 8,
      "paddingHorizontal": 16,
    }
  }
>
  <Text
    style={
      Object {
        "color": "hsl(0, 0%, 0%)",
        "fontSize": 14,
        "fontWeight": "600",
        "letterSpacing": 0.75,
        "opacity": 1,
        "textAlign": "center",
      }
    }
  >
    HELLO
  </Text>
</View>
`;

exports[`renders disable 1`] = `
<View
  accessible={true}
  focusable={true}
  onClick={[Function]}
  onResponderGrant={[Function]}
  onResponderMove={[Function]}
  onResponderRelease={[Function]}
  onResponderTerminate={[Function]}
  onResponderTerminationRequest={[Function]}
  onStartShouldSetResponder={[Function]}
  style={
    Object {
      "alignItems": "center",
      "alignSelf": "center",
      "backgroundColor": "transparent",
      "borderColor": "transparent",
      "borderRadius": 4,
      "borderWidth": 1,
      "flexDirection": "row",
      "justifyContent": "center",
      "opacity": 1,
      "padding": 8,
      "paddingHorizontal": 16,
    }
  }
>
  <Text
    style={
      Object {
        "color": "hsla(0, 0%, 0%, 0.38)",
        "fontSize": 14,
        "fontWeight": "600",
        "letterSpacing": 0.75,
        "opacity": 1,
        "textAlign": "center",
      }
    }
  >
    HELLO
  </Text>
</View>
`;
